framework/auth/views.py

```diff
@@ -198,88 +198,155 @@ def forgot_password_post(auth, **kwargs):
     return {}
 
 
-@collect_auth
-def auth_login(auth, **kwargs):
+def login_and_register_handler(auth, login=True, campaign=None, service_url=None, logout=None, **kwargs):
+    """
+    Non-view helper to handle `login` and `register` requests.
+
+    :param auth: the auth context
+    :param login: login or register
+    :param campaign: `institution` or `prereg`
+    :param service_url: the service url for CAS or redirect url for OSF
+    :param logout: used only for `claim_user_registered`
+    :param kwargs: None
+    :return: data object that contains actions for `auth_register` and `auth_login`
+    :raises: http.BAD_REQUEST
     """
-    This view serves as the entry point for OSF login and campaign login.
-    HTTP Method: GET
 
-        GET '/login/' without any query parameter:
-            redirect to CAS login page with dashboard as target service
+    if not service_url:
+        service_url = web_url_for('dashboard', _absolute=True)
 
-        GET '/login/?logout=true
-            log user out and redirect to CAS login page with redirect_url or next_url as target service
+    if not validate_service_url(service_url):
+        raise HTTPError(http.BAD_REQUEST)
 
-        GET '/login/?campaign=instituion:
-            if user is logged in, redirect to 'dashboard'
-            show institution login
+    data = {
+        'status_code': http.FOUND if login else http.OK,
+        'redirect_url': service_url,
+        'campaign': None,
+        'must_login_warning': False,
+        'service_url': service_url,
+    }
+
+    if campaign:
+        if validate_campaign(campaign):
+            if campaign == 'prereg':
+                # `GET /login?campaign=prereg`
+                if login:
+                    service_url = data['redirect_url'] = web_url_for('auth_register', campaign='prereg')
+                # `GET /register?campaign=prereg`
+                else:
+                    data['campaign'] = campaign
+                    data['must_login_warning'] = True
+                    service_url = data['service_url'] = campaigns.campaign_url_for(campaign)
+            # GET `/register` or '/login` with `campaign=institution`
+            elif campaign == 'institution':
```

Comment:
(<<< [previous](https://github.com/CenterForOpenScience/osf.io/pull/6082/files#r74649603) ) ... this should be the first conditional (`if/else` instead of `if/elif/else`)